core/src/test/scala/unit/kafka/utils/TestUtils.scala

```diff
@@ -1127,6 +1127,23 @@ object TestUtils extends Logging {
       throw new IllegalStateException(s"Cannot get topic: $topic, partition: $partition in server metadata cache"))
   }
 
+  /**
+   * Wait until the kraft broker metadata have caught up to the controller
+   */
+  def waitForKRaftBrokerMetadataCatchupController(
+      brokers: Seq[KafkaBroker],
+      controllerServer: ControllerServer,
+      msg: String = "Timeout waiting for controller metadata propagating to brokers"
+  ): Unit = {
+    TestUtils.waitUntilTrue(
+      () => {
+        brokers.forall { broker =>
+          val metadataOffset = broker.metadataCache.asInstanceOf[KRaftMetadataCache].currentImage().highestOffsetAndEpoch().offset
```

Comment:
Wonder if we should look at the last published offset?

Reply:
Yes, the image is reset when we start publishing so it's still not very accurate, I added a `MetadataPublisher.publishedOffset` method to accomplish this.